### PR TITLE
very basic definitions

### PR DESCRIPTION
**Pull request by mapsam**
Adding some glossary definitions for the first few entries. Converted all of the lists into headers so we can link to them per [comment](https://github.com/mapbox/cpp/issues/2#issuecomment-254291416).

cc @springmeyer @GretaCB @mikemorris

### CODE REVIEW

**Comment by mapsam** [comment]
Woops, this is a PR into the node-addons branch ... but that's probably okay.

**Comment by mikemorris** [comment]
screen shot 2016-10-17 at 9 31 35 pm

screen shot 2016-10-17 at 9 31 20 pm

**Comment by mapsam** [comment]
Ha, gosh darnit @mikemorris I did that and now it exposed my poor git techniques and brought in all of my commits from the other branch 😄

**Comment by mapsam** [comment]
Rebased, all good.

File: glossary.md [hunk 1]
```diff
@@ -1,49 +1,95 @@
...
- - How is C/C++ used in operating systems like Linux and OS X
\ No newline at end of file
+# Building C++
+
+#### development toolchain
+
+The toolset used to generate C++ code into [executables](#executable) for a computer. Typically consists of a [compiler](#compiler) and [linker](#linker). 
+
+All C++ code falls into the ["ahead-of-time" compilation](https://en.wikipedia.org/wiki/Ahead-of-time_compilation) category, which means code is compiled and linked separately from execution, unlike  ["just-in-time"](https://en.wikipedia.org/wiki/Just-in-time_compilation) compilation languages, like JavaScript.
+
```
Comment (springmeyer) [inline glossary.md:8]: Nice links!

File: glossary.md [hunk 2]
```diff
@@ -1,49 +1,95 @@
...
+#### compiler
+
+A program that converts a C++ source code into machine code (binary) files readable by a computer. Compilers turn `.cpp` code into `.o` [object files](#object-file). 
+
+Example compilers are [clang](#clang) or [gcc](#gcc).
```
Comment (springmeyer) [inline glossary.md]: I would say `Example open source compilers are`. There are many compilers and some that purport to be better for more $ (like https://software.intel.com/en-us/c-compilers which may have optimizations specifically for intel chips). But we are in an era now where the open source ones are likely just as good if not better for general purpose computing.

File: glossary.md [hunk 3]
```diff
@@ -1,49 +1,95 @@
...
+
+#### linker
+
+A program that combines object files into a single [executable](#executable) or [library](#library). Typically a step enacted by the [compiler](#compiler).
+
```
Comment (springmeyer) [inline glossary.md:18]: 👍

File: glossary.md [hunk 4]
```diff
@@ -1,49 +1,95 @@
...
+#### object file
+
+An object file contains symbols, compiled code, external symbols, and other static data. If you are compiling three C++ files – `one.cpp`, `two.cpp`, `three.cpp` – they will be compiled into three object files – `one.o`, `two.o`, `three.o`. These can subsequently be combined by a [linker](#linker) and turned into an [executable](#executable).
+
```
Comment (springmeyer) [inline glossary.md:22]: `and turned into an [executable](#executable).` should be `and turned into a [library](#library) or [executable](#executable).`

File: glossary.md [hunk 5]
```diff
@@ -1,49 +1,95 @@
...
+#### executable
+
+A binary file and/or program that can be run by a computer. This is the outcome of [compiling](#compiler) and [linking](#linker).
+
```
Comment (springmeyer) [inline glossary.md:26]: An executable specifically has a `main()` function, aka `entry point`. A library does not.

File: glossary.md [hunk 6]
```diff
@@ -1,49 +1,95 @@
...
+#### posix
+
+A set of standards for maintaining compatability between Unix-like operating systems. *HELP: not sure how this can be explained from our workflow perspective at Mapbox*
+
```
Comment (springmeyer) [inline glossary.md:30]: Generally `posix` is synonymous with `unix` when it comes to what system APIs exist in C/C++ code to interface with the operating system. With minor exceptions POSIX APIs work the same on linux and osx.

File: glossary.md [hunk 7]
```diff
@@ -1,49 +1,95 @@
...
+#### library
+
+A set of reusable C++ code that can be shared across projects. Libraries can be organized very differently, but typically contain a set of header files (static library) or pre-compiled binaries (dynamic/shared library).
+
```
Comment (springmeyer) [inline glossary.md:36]: - libraries lack a `main` function and instead provide an interface (functions, classes, etc) that can be called by other C/C++ libraries or executables.
- Libraries must provide at least one header file to allow calling applications to know the definitions of the interfaces it provides.
- Libraries can either be compiled at `static archives` (aka static library) or as a shared (aka dynamic library).

File: glossary.md [hunk 8]
```diff
@@ -1,49 +1,95 @@
...
+#### dynamic/shared library
+
+A shared library is only required at runtime, meaning it has already been [compiled](#compiler) into machine-readable code and is just required alongside your code when it executes.
+
+#### static library
+
+A static library is required at compile time, meaning all of the code in your include directory will be [compiled](#compiler) and [linked](#linker) into your final [executable](#executable). These libraries typically consist of [header files](#header) that can be included directly into source code, with the `#include "library.h"` syntax.
+
```
Comment (springmeyer) [inline glossary.md:44]: Edited: can you replace with:

> A static library is not required at runtime. Instead when a precompiled static library is linked by a calling application all of its object code is linked into that calling applications executable. This step is often called "static linking". The source code of a static library looks identical to a shared library (both contain headers and .cpp files). The only difference is the way it is built.

**Review by springmeyer — APPROVED** [review]

File: glossary.md (hunk 2 above)
Comment (mapsam) [reply glossary.md]: If someone wants to go to Best Buy and purchase a compiler, let em!

File: glossary.md (hunk 6 above)
Comment (mapsam) [reply glossary.md:30]: I'm going to copy this verbatim. Okay @springmeyer?

**Comment by mapsam** [comment]
Thanks for the comments @springmeyer! I've gone through all of them and updated in https://github.com/mapbox/cpp/pull/14/commits/754ed1bf8384bbe31660017ea42091dbd10ac5d9

**Comment by springmeyer** [comment]
Great! Let's merge this and keep forging ahead.

**Pull request merged**